price-alert-script/price-alert-script.py

```diff
@@ -31,34 +34,99 @@ def get_database_connection() -> connection:
+    for row in latest_prices:
+        row['updated_at'] = datetime.fromtimestamp(
+            row['updated_at'].timestamp()).strftime('%Y-%m-%d %H:%M:%S.%f')
+        row['price'] = float(row['price'])
 
+    # sort the dictionaries so they're paired by product id
+    latest_prices = sorted(latest_prices, key=lambda x: x['product_id'])
+    grouped_by_product_id = {key: list(group) for key, group in groupby(
+        latest_prices, key=lambda x: x['product_id'])}
 
-def compare_latest_two_prices():
-    """Compare values of price for last two  entries of each product"""
-    pass
+    return grouped_by_product_id
+
+
+def compare_latest_two_prices(latest_price_entries: dict, product_id_of_interest: int) -> bool:
+    """
+    Compare values of price for last two  entries of each product.
+    """
+    # make function take in product id and return true false for just one product
+
+    # takes in the result of prev function and compares the last two entries for a product
+    # outputs bool. If price increases, return true
+
+    for key, value in latest_price_entries.items():
+        if key == product_id_of_interest:
+            products_of_interest = value
+            # this will only return one list because they have all been grouped previously
+    # conditions for if the price has increased or stayed same. True if theres an increase.
+    if products_of_interest[-1]['price'] < products_of_interest[-2]['price']:
+        return True
+
+    else:
+        return False
+    # condense this ^^
 
 
 def get_user_emails(rds_conn: connection):
-    """Query RDS user table for User emails"""
-    pass
+    """
+    Query RDS user table for User emails.
+    """
+    list_of_lists = []
+    # ^needs a better name. Holds all the values from the point at the bottom of functn
+    cur = rds_conn.cursor(cursor_factory=extras.RealDictCursor)
+
+    # Find all subscriptions
+    cur.execute("SELECT * FROM subscriptions;")
+    # makes it a list of dicts
+    all_subscriptions = [dict(row) for row in cur.fetchall()]
+
+    for subscription in all_subscriptions:
+        user_id = subscription['user_id']
+        cur.execute("SELECT email FROM users WHERE user_id = (%s);", (user_id,))
+        user_email = [dict(row) for row in cur.fetchall()][0]['email']
+        product_id = subscription['product_id']
+
+        # get product name
+        cur.execute(
+            "SELECT product_name FROM products WHERE product_id = (%s);", (product_id,))
+        product_name = [dict(row) for row in cur.fetchall()][0]['product_name']
+
+        is_discounted = compare_latest_two_prices(
+            get_prices_of_latest_pair_of_products(rds_conn), product_id)
+
+        list_of_lists.append(
+            [user_id, user_email, product_id, product_name, is_discounted])
+
+    # return a dict of the user email, product name/id, true or false for if its been updated
+
+    return list_of_lists  # make this a dict instead? Will me more consistent indexing
```

Comment:
For this could make the list_of_lists = {} then rather than a list and instead of "list_of_lists.append()" use list_of_lists[""e.g. user_id"] = [user_id, user_email, product_id, product_name, is_discounted]. 

The dict key can change to what you need it to be. 

list_of_lists could be renamed to all_user_prods ?